rcl/test/rcl/test_graph.cpp

```diff
@@ -310,7 +313,218 @@ check_graph_state(
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  size_t number_of_tries = 20;
+  rcutils_string_array_t node_names = rcutils_get_zero_initialized_string_array();
+  rcutils_string_array_t node_namespaces = rcutils_get_zero_initialized_string_array();
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
+                                      ret = rcutils_string_array_fini(&node_names);
+                                      ASSERT_EQ(RCUTILS_RET_OK, ret);
+                                      ret = rcutils_string_array_fini(&node_namespaces);
+                                      ASSERT_EQ(RCUTILS_RET_OK, ret);
+                                  });
+  size_t attempts = 0;
+  size_t max_attempts = 4;
+  while (node_names.size < 3) {
+    std::this_thread::sleep_for(std::chrono::seconds(1));
+    ret = rcl_get_node_names(remote_node_ptr, allocator, &node_names, &node_namespaces);
+    attempts++;
+    ASSERT_LE(attempts, max_attempts) << "Unable to attain all required nodes";
+  }
+
+  check_graph_state(
+    this->node_ptr,
+    this->wait_set_ptr,
+    graph_guard_condition,
+    topic_name,
+    1,  // expected publishers on topic
+    2,  // expected subscribers on topic
+    true,  // topic expected in graph
+    number_of_tries);  // number of retries
+  check_graph_state(
+    remote_node_ptr,
+    this->wait_set_ptr,
+    remote_graph_guard_condition,
+    topic_name,
+    1,  // expected publishers on topic
+    2,  // expected subscribers on topic
+    true,  // topic expected in graph
+    number_of_tries);  // number of retries
+  GetTopicsFunc sub_func = std::bind(rcl_get_subscriber_names_and_types_by_node,
+                                     std::placeholders::_1,
+                                     &allocator,
+                                     false,
+                                     std::placeholders::_2,
+                                     "/",
+                                     std::placeholders::_3);
+  GetTopicsFunc pub_func = std::bind(rcl_get_publisher_names_and_types_by_node,
+                                     std::placeholders::_1,
+                                     &allocator,
+                                     false,
+                                     std::placeholders::_2,
+                                     "/",
+                                     std::placeholders::_3);
+  GetTopicsFunc service_func = std::bind(rcl_get_service_names_and_types_by_node,
+                                     std::placeholders::_1,
+                                     &allocator,
+                                     std::placeholders::_2,
+                                     "/",
+                                     std::placeholders::_3);
+  std::array<rcl_node_t*, 2> node_array {remote_node_ptr, node_ptr};
+
+  // verify each node contains the same node graph.
```

Comment:
I'm not sure why we need to create a second node for this test. IIUC, we can sufficiently test the new functions with one node.

Reply:
Participant discovery only triggers for remote nodes, whereas endpoint discovery triggers for all nodes (including the current node). Two nodes are required to test the functionality where inspecting the current node versus remote nodes.